packages/@aws-cdk/aws-codepipeline-actions/lib/ecs/deploy-action.ts

```diff
@@ -0,0 +1,105 @@
+   * If you use this property, it's assumed the file is called 'imagedefinitions.json'.
+   * If your build uses a different file, leave this property empty,
+   * and use the `imageFile` property instead.
+   *
+   * @default - one of this property, or `imageFile`, is required
+   * @see https://docs.aws.amazon.com/codepipeline/latest/userguide/pipelines-create.html#pipelines-create-image-definitions
+   */
+  readonly inputArtifact?: codepipeline.Artifact;
+
+  /**
+   * The name of the JSON image definitions file to use for deployments.
+   * The JSON file is a list of objects,
+   * each with 2 keys: `name` is the name of the container in the Task Definition,
+   * and `imageUri` is the Docker image URI you want to update your service with.
+   * Use this property if you want to use a different name for this file than the default 'imagedefinitions.json'.
+   * If you use this property, you don't need to specify the `inputArtifact` property.
+   *
+   * @default - one of this property, or `inputArtifact`, is required
+   * @see https://docs.aws.amazon.com/codepipeline/latest/userguide/pipelines-create.html#pipelines-create-image-definitions
+   */
+  readonly imageFile?: codepipeline.ArtifactPath;
+
+  /**
+   * The ECS Service to deploy.
+   */
+  readonly service: ecs.BaseService;
+}
+
+/**
+ * CodePipeline Action to deploy an ECS Service.
+ */
+export class EcsDeployAction extends codepipeline.DeployAction {
+  constructor(props: EcsDeployActionProps) {
+    super({
+      ...props,
+      inputArtifact: determineInputArtifact(props),
+      provider: 'ECS',
+      artifactBounds: {
+        minInputs: 1,
+        maxInputs: 1,
+        minOutputs: 0,
+        maxOutputs: 0,
+      },
+      configuration: {
+        ClusterName: props.service.clusterName,
+        ServiceName: props.service.serviceName,
+        FileName: props.imageFile ? props.imageFile.fileName : undefined,
+      },
+    });
+  }
+
+  protected bind(info: codepipeline.ActionBind): void {
+    // permissions based on CodePipeline documentation:
+    // https://docs.aws.amazon.com/codepipeline/latest/userguide/how-to-custom-role.html#how-to-update-role-new-services
+    info.role.addToPolicy(new iam.PolicyStatement()
+      .addActions(
+        'ecs:DescribeServices',
+        'ecs:DescribeTaskDefinition',
+        'ecs:DescribeTasks',
+        'ecs:ListTasks',
+        'ecs:RegisterTaskDefinition',
+        'ecs:UpdateService',
+      )
+      .addAllResources());
+
+    info.role.addToPolicy(new iam.PolicyStatement()
+      .addActions(
+        'iam:PassRole',
+      )
+      .addAllResources()
+      .addCondition('StringEqualsIfExists', {
+        'iam:PassedToService': [
+          'ec2.amazonaws.com',
+          'ecs-tasks.amazonaws.com',
+        ],
+      }));
+  }
+}
+
+function determineInputArtifact(props: EcsDeployActionProps): codepipeline.Artifact {
+  if (props.imageFile) {
+    if (props.inputArtifact) {
```

Comment:
Start with:
```ts
if (!!props.imageFile === !!props.inputArtifact) {
  throw new Error(`Exactly one of impageFile or inputArtifact must be provided, but ${!!props.imageFile ? 'both' : 'none'} was provided.`);
}
```

Then you don't have to nest so deep anymore.

Reply:
I don't really like that. You have 4 exits from this method, and each has is its own branch, which seems vary natural and readable to me. The nesting level is only 2, which I don't consider a problem.

Your version uses an (in my opinion) non-obvious JS idiom, and forces me to use `!` in the exits that do actually return a value, as the compiler won't be able to tell which of the properties is definitely not `undefined`. I don't think those things are worth doing to save one level of nesting.